File: core/prompt/world_info_scanner.py
"""
core.prompt.world_info_scanner — World Info 触发扫描器

扫描聊天历史 + 角色/用户设定，激活匹配关键词的 WI 条目，
并按注入位置分流。参考 SillyTavern 的 WIGlobalScanData 机制。
"""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass, field
from typing import List

from models.world_info import WIPosition, WorldInfoBook, WorldInfoEntry

logger = logging.getLogger(__name__)


@dataclass
class ActivatedEntry:
    """一个被激活的 WI 条目"""

    entry: WorldInfoEntry
    book_name: str


@dataclass
class ScanResult:
    """扫描结果 — 按注入位置分组"""

    before_char: List[ActivatedEntry] = field(default_factory=list)
    after_char: List[ActivatedEntry] = field(default_factory=list)
    before_examples: List[ActivatedEntry] = field(default_factory=list)
    after_examples: List[ActivatedEntry] = field(default_factory=list)
    at_depth: List[ActivatedEntry] = field(default_factory=list)
    system_top: List[ActivatedEntry] = field(default_factory=list)
    system_bottom: List[ActivatedEntry] = field(default_factory=list)

    def get_by_position(self, position: WIPosition) -> List[ActivatedEntry]:
        """按位置获取已激活条目"""
        mapping = {
            WIPosition.BEFORE_CHAR: self.before_char,
            WIPosition.AFTER_CHAR: self.after_char,
            WIPosition.BEFORE_EXAMPLES: self.before_examples,
            WIPosition.AFTER_EXAMPLES: self.after_examples,
            WIPosition.AT_DEPTH: self.at_depth,
            WIPosition.SYSTEM_TOP: self.system_top,
            WIPosition.SYSTEM_BOTTOM: self.system_bottom,
        }
        return mapping.get(position, [])

    @property
    def total_activated(self) -> int:
        return sum(
            len(lst)
            for lst in [
                self.before_char,
                self.after_char,
                self.before_examples,
                self.after_examples,
                self.at_depth,
                self.system_top,
                self.system_bottom,
            ]
        )


class WorldInfoScanner:
    """
    WI 触发扫描器。

    扫描源（参考 SillyTavern 的 WIGlobalScanData）：
    - 聊天历史全文
    - character.description / personality / scenario
    - persona.description
    - creator_notes
    """

    def scan(
        self,
        books: List[WorldInfoBook],
        scan_text: str,
    ) -> ScanResult:
        """
        扫描给定文本，激活匹配的 WI 条目并按位置分组。

        Args:
            books: 要扫描的知识库列表
            scan_text: 拼接后的扫描语料
        """
        result = ScanResult()
        scan_lower = scan_text.lower()

        for book in books:
            if not book.enabled:
                continue
            for entry in book.entries:
                if not entry.enabled:
                    continue
                if entry.constant or self._matches(entry, scan_lower):
                    activated = ActivatedEntry(
                        entry=entry, book_name=book.name
                    )
                    self._route(activated, result)

        # 每个分组内部按 order 排序
        for group in [
            result.before_char,
            result.after_char,
            result.before_examples,
            result.after_examples,
            result.at_depth,
            result.system_top,
            result.system_bottom,
        ]:
            group.sort(key=lambda a: a.entry.order)

        if result.total_activated > 0:
            logger.info(
                f"World Info 扫描完成: {result.total_activated} 条目被激活"
            )

        return result

    def _matches(self, entry: WorldInfoEntry, scan_text: str) -> bool:
        """检查条目是否被触发"""
        # 主关键词：任一匹配即触发
        primary_hit = any(
            self._keyword_in_text(key, scan_text) for key in entry.keys
        )
        if not primary_hit:
            return False

        # 如果没有二级关键词，主关键词匹配即可
        if not entry.secondary_keys:
            return True

        # 二级关键词逻辑
        if entry.selective_logic == "AND":
            return any(
                self._keyword_in_text(key, scan_text)
                for key in entry.secondary_keys
            )
        elif entry.selective_logic == "NOT":
            return not any(
                self._keyword_in_text(key, scan_text)
                for key in entry.secondary_keys
            )

        return True

    def _keyword_in_text(self, keyword: str, text: str) -> bool:
        """关键词匹配（大小写不敏感，支持中文）"""
        keyword_lower = keyword.lower().strip()
        if not keyword_lower:
            return False
        # 用 re.escape 安全匹配特殊字符
        return bool(re.search(re.escape(keyword_lower), text))

    def _route(self, activated: ActivatedEntry, result: ScanResult) -> None:
        """按配置的注入位置分流到对应的分组"""
        position = activated.entry.position
        if position == WIPosition.BEFORE_CHAR:
            result.before_char.append(activated)
        elif position == WIPosition.AFTER_CHAR:
            result.after_char.append(activated)
        elif position == WIPosition.BEFORE_EXAMPLES:
            result.before_examples.append(activated)
        elif position == WIPosition.AFTER_EXAMPLES:
            result.after_examples.append(activated)
        elif position == WIPosition.AT_DEPTH:
            result.at_depth.append(activated)
        elif position == WIPosition.SYSTEM_TOP:
            result.system_top.append(activated)
        elif position == WIPosition.SYSTEM_BOTTOM:
            result.system_bottom.append(activated)

File: backend/src/services/world-info-scanner.ts
import type {
  Character,
  Message,
  WorldInfoBook,
  WorldInfoEntry,
} from "@ai-party/shared";
import {
  evaluateVariableConditions,
  type VariableConditionContext,
} from "./variable-conditions";

export interface ActivatedEntry {
  entry: WorldInfoEntry;
  bookName: string;
}

export interface ScanResult {
  before_char: ActivatedEntry[];
  after_char: ActivatedEntry[];
  before_examples: ActivatedEntry[];
  after_examples: ActivatedEntry[];
  at_depth: ActivatedEntry[];
  system_top: ActivatedEntry[];
  system_bottom: ActivatedEntry[];
}

export interface WorldInfoScanOptions {
  variableContext?: VariableConditionContext;
}

export function createEmptyScanResult(): ScanResult {
  return {
    before_char: [],
    after_char: [],
    before_examples: [],
    after_examples: [],
    at_depth: [],
    system_top: [],
    system_bottom: [],
  };
}

export function getScanResultByPosition(
  result: ScanResult,
  position: WorldInfoEntry["position"],
): ActivatedEntry[] {
  switch (position) {
    case "before_char":
      return result.before_char;
    case "after_char":
      return result.after_char;
    case "before_examples":
      return result.before_examples;
    case "after_examples":
      return result.after_examples;
    case "at_depth":
      return result.at_depth;
    case "system_top":
      return result.system_top;
    case "system_bottom":
      return result.system_bottom;
    default:
      return [];
  }
}

export function countActivatedEntries(result: ScanResult): number {
  return (
    result.before_char.length +
    result.after_char.length +
    result.before_examples.length +
    result.after_examples.length +
    result.at_depth.length +
    result.system_top.length +
    result.system_bottom.length
  );
}

export function buildWorldInfoScanText(
  character: Character,
  chatHistory: Message[],
  personaDescription = "",
): string {
  const parts: string[] = [
    character.personality,
    character.background,
    character.description || "",
    character.scenario || "",
    character.creator_notes || "",
  ];

  if (personaDescription.trim()) {
    parts.push(personaDescription.trim());
  }

  for (const message of chatHistory.slice(-30)) {
    if (!message.is_system) {
      parts.push(message.content);
    }
  }

  return parts.filter(Boolean).join("\n");
}

function sortByOrder(entries: ActivatedEntry[]): void {
  entries.sort((left, right) => left.entry.order - right.entry.order);
}

function routeActivatedEntry(activated: ActivatedEntry, result: ScanResult): void {
  switch (activated.entry.position) {
    case "before_char":
      result.before_char.push(activated);
      break;
    case "after_char":
      result.after_char.push(activated);
      break;
    case "before_examples":
      result.before_examples.push(activated);
      break;
    case "after_examples":
      result.after_examples.push(activated);
      break;
    case "at_depth":
      result.at_depth.push(activated);
      break;
    case "system_top":
      result.system_top.push(activated);
      break;
    case "system_bottom":
      result.system_bottom.push(activated);
      break;
    default:
      break;
  }
}

function keywordInText(keyword: string, text: string): boolean {
  const normalized = keyword.toLowerCase().trim();
  if (!normalized) {
    return false;
  }

  const escaped = normalized.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return new RegExp(escaped, "i").test(text);
}

function matchesEntry(entry: WorldInfoEntry, scanTextLower: string): boolean {
  const primaryHit = entry.keys.some((key) => keywordInText(key, scanTextLower));
  if (!primaryHit) {
    return false;
  }

  if (entry.secondary_keys.length === 0) {
    return true;
  }

  if (entry.selective_logic === "AND") {
    return entry.secondary_keys.some((key) => keywordInText(key, scanTextLower));
  }

  if (entry.selective_logic === "NOT") {
    return !entry.secondary_keys.some((key) => keywordInText(key, scanTextLower));
  }

  return true;
}

export class WorldInfoScanner {
  scan(books: WorldInfoBook[], scanText: string, options: WorldInfoScanOptions = {}): ScanResult {
    const result = createEmptyScanResult();
    const scanLower = scanText.toLowerCase();
    const variableContext = options.variableContext || { room: {}, global: {} };

    for (const book of books) {
      if (!book.enabled) {
        continue;
      }

      for (const entry of book.entries) {
        if (!entry.enabled) {
          continue;
        }

        const matchesTrigger = entry.constant || matchesEntry(entry, scanLower);
        const matchesConditions = evaluateVariableConditions(
          entry.conditions,
          entry.condition_logic,
          variableContext,
        );

        if (matchesTrigger && matchesConditions) {
          routeActivatedEntry({ entry, bookName: book.name }, result);
        }
      }
    }

    for (const group of [
      result.before_char,
      result.after_char,
      result.before_examples,
      result.after_examples,
      result.at_depth,
      result.system_top,
      result.system_bottom,
    ]) {
      sortByOrder(group);
    }

    return result;
  }
}
